mlir/include/mlir/Interfaces/LoopLikeInterface.td

```diff
@@ -20,6 +20,19 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
     Contains helper functions to query properties and perform transformations
     of a loop. Operations that implement this interface will be considered by
     loop-invariant code motion.
+
+    Loop-carried variables can be exposed through this interface. There are
+    3 components to a loop-carried variable.
+    - The "region iter_arg" is the block argument of the entry block that
+      represents the loop-carried variable in each iteration.
+    - The "init value" is an operand of the loop op that serves as the initial
+      region iter_arg value for the first iteration (if any).
+    - The "yielded" value is the value that is forwarded from one iteration to
+      serve as the region iter_arg of the next iteration.
+
+    If one of the respective interface methods is implemented, so must the other
+    two. The interface verifier ensures that the number of types of the region
```

Comment:
(sorry, I looked at this last week but apparently forgot to submit and only noticed now)

Reply:
That's correct. The bbArg of this op is also a bit different from loop iter_args because it is never a yielded value from another iteration. It is always the "init" operand (and accessing parts that were written by another iteration is undefined).

Reply:
I see, but does it make sense to ban Ops like `ForallOp` from implementing `getInits()`? When looking at the `LoopLikeOp` from the outside, it takes an init operand, then works on it (either by passing it along the iterations or in parallel) and then produces a result.

Reply:
The problem is if we allow `getInits` without the other two functions, we cannot verify certain invariants anymore (e.g., matching number of yielded values/init operands and their types). And future transformations (I'm planning to add one for "subset hoisting") that operate on arbitrary  init-iter_arg-yield style loops will need special casing.

Note, `scf.forall` can still implement the `getDpsInitsMutable()` method, which basically says that the init operand is tied to the OpResult. Maybe that is the abstraction that you are actually interested in?

Reply:
> we cannot verify certain invariants anymore

Ah, makes sense.

> Note, `scf.forall` can still implement the `getDpsInitsMutable()` method

But what about other "looplike" Ops that don't implement the Dps Interface like `scf.parallel`? I think being able to query the tensor/buffer a `LoopLikeOpInterface` operates on is very useful for loop nest analysis - it tells us that the respective buffer is accessed "tripcount" times, where sometimes it's less important whether that is done sequentially or in parallel in the actual Operation.
So I'm not arguing against this change, it's more about not excluding `forall` and `parallel` from providing buffer information through the Interface

Reply:
I could change rename `getInits`, `getYieldedValues`, `getRegionIterArgs` to `getLoopInits`, `getLoopYieldedValues`, `getLoopRegionIterArgs` and change the return value to `std::optional<...>`. Then a subset of the methods can be implemented. The renaming is needed because ops such as `scf::WhileOp` already have an auto-generated `getInits` function (that matches with the implementation that the interface expects).

Reply:
> The renaming is needed because ops such as `scf::WhileOp` already have an auto-generated `getInits` function (that matches with the implementation that the interface expects).
> 
> This means we would have both `getInits` and `getLoopInits` etc. on most loop ops.

Ah, I see now why this is more complicated than just allowing the "non-sequentially-iterated loops" to implement `getInits()`.
Sorry - I didn't want to stall this PR, I'm fine with going ahead with it as it is now and trying to adapt the interface in the future to fit the different cases better. Maybe this is also a case where we are trying to square a circle by trying to cover both parallel-like and sequential-like loops perfectly with one interface - maybe this could be solved by interface inheritance in the future.